common/src/types/v0/store/node.rs

```diff
@@ -1,31 +1,104 @@
 //! Definition of node types that can be saved to the persistent store.
 
-use crate::types::v0::{
-    openapi::models,
-    store::{
-        definitions::{ObjectKey, StorableObject, StorableObjectType},
-        ResourceUuid,
+use crate::{
+    types::v0::{
+        openapi::models,
+        store::{
+            definitions::{ObjectKey, StorableObject, StorableObjectType},
+            ResourceUuid,
+        },
+        transport::{self, NodeId},
     },
-    transport::{self, NodeId},
+    IntoOption,
 };
 use serde::{Deserialize, Serialize};
 use std::collections::HashMap;
 
+/// Generic labels associated with the node
 pub type NodeLabels = HashMap<String, String>;
 
+/// Data relating to the cordoning of a node
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
+pub struct CordonedState {
+    /// labels used to enforce a node cordon
+    pub cordonlabels: Vec<String>,
+}
+
+impl CordonedState {
+    pub fn new(cordonlabels: Vec<String>) -> Self {
+        Self { cordonlabels }
+    }
+    pub fn add_cordon_label(&mut self, label: String) {
+        self.cordonlabels.insert(0, label);
+    }
+    pub fn remove_cordon_label(&mut self, label: String) {
+        if let Some(index) = self.cordonlabels.iter().position(|l| l == &label.clone()) {
+            self.cordonlabels.remove(index);
+        }
+    }
+}
+
+/// Data relating to a draining or drained node, including non-drain cordon labels
+#[derive(Clone, Serialize, Deserialize, Debug, PartialEq)]
+pub struct DrainState {
+    /// labels used to enforce a node cordon
+    pub cordonlabels: Vec<String>,
+    /// labels used to cordon a node specifically for a drain
+    pub drainlabels: Vec<String>,
```

Comment:
Why do we need to distinguish the labels again? Once we're in drain, we can only move to down to uncordon and not cordoned?